app/services/users/omniauth.rb

```diff
@@ -6,31 +6,37 @@ def initialize(auth)
   end
 
   def get
-    user
+    check_auth
+    @user
   end
 
   private
 
   delegate :provider, :uid, :info, to: :auth
   delegate :email, to: :info
 
-  def user
-    @_user ||= find_user || create_user
+  def find_authorization(user)
+    user.authorizations.where(provider: provider, uid: uid).first
   end
 
-  def find_user
-    User.where(provider: provider, uid: uid, email: email).first
+  def create_authorization(user)
+    user.authorizations.create(provider: provider, uid: uid)
+  end
+
+  def initiate_user
+    @user = User.where(email: email).first || create_user
+  end
+
+  def check_auth
+    initiate_user
+    create_authorization(@user) if find_authorization(@user).blank?
```

Comment:
@sinpai [find_or_create_by](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_create_by) can be used

Reply:
Yes, I've looked at it, but construction with block didn't get in my head) Thanks.